crates/sui-core/src/authority/shared_object_version_manager.rs

```diff
@@ -153,30 +161,60 @@ async fn get_or_init_versions(
+    cancelled_txns: &BTreeMap<TransactionDigest, CancelConsensusCertificateReason>,
 ) -> Vec<(ObjectID, SequenceNumber)> {
     let tx_digest = cert.digest();
 
+    // Check if the transaction is cancelled due to congestion.
+    let cancellation_info: Option<HashSet<_>> =
+        if let Some(CancelConsensusCertificateReason::CongestionOnObjects(congested_objects)) =
+            cancelled_txns.get(tx_digest)
+        {
+            Some(congested_objects.clone().into_iter().collect())
+        } else {
+            None
+        };
+    let txn_cancelled = cancellation_info.is_some();
+
     // Make an iterator to update the locks of the transaction's shared objects.
     let shared_input_objects: Vec<_> = cert.shared_input_objects().collect();
 
-    let mut input_object_keys =
-        transaction_input_object_keys(cert).expect("Transaction input should have been verified");
+    let mut input_object_keys = transaction_non_shared_input_object_keys(cert)
+        .expect("Transaction input should have been verified");
     let mut assigned_versions = Vec::with_capacity(shared_input_objects.len());
     let mut is_mutable_input = Vec::with_capacity(shared_input_objects.len());
     // Record receiving object versions towards the shared version computation.
     let receiving_object_keys = transaction_receiving_object_keys(cert);
     input_object_keys.extend(receiving_object_keys);
 
-    for (SharedInputObject { id, mutable, .. }, version) in shared_input_objects
-        .iter()
-        .map(|obj| (obj, *shared_input_next_versions.get(&obj.id()).unwrap()))
-    {
-        assigned_versions.push((*id, version));
-        input_object_keys.push(ObjectKey(*id, version));
-        is_mutable_input.push(*mutable);
+    if let Some(congested_objects) = cancellation_info {
+        // For cancelled transaction due to congestion, assign special versions to all shared objects.
+        // Note that new lamport version does not depend on any shared objects.
+        for SharedInputObject { id, .. } in shared_input_objects.iter() {
+            let assigned_version = if congested_objects.contains(id) {
+                SequenceNumber::CONGESTED
+            } else {
+                SequenceNumber::CANCELLED_READ
+            };
+            assigned_versions.push((*id, assigned_version));
+            is_mutable_input.push(false);
+        }
+    } else {
+        for (SharedInputObject { id, mutable, .. }, assigned_version) in shared_input_objects
+            .iter()
+            .map(|obj| (obj, *shared_input_next_versions.get(&obj.id()).unwrap()))
+        {
+            assigned_versions.push((*id, assigned_version));
+            input_object_keys.push(ObjectKey(*id, assigned_version));
+            is_mutable_input.push(*mutable);
+        }
     }
 
     let next_version = SequenceNumber::lamport_increment(input_object_keys.iter().map(|obj| obj.1));
+    assert!(
+        next_version < SequenceNumber::MAX,
```

Comment:
might be worth adding an `assert_valid()` method to SequenceNumber?

Reply:
Added is_valid(), and can be used in the assertion.